src/commonMain/kotlin/fr/acinq/lightning/payment/OutgoingPaymentFailure.kt

```diff
@@ -34,26 +76,73 @@ data class OutgoingPaymentFailure(val reason: FinalFailure, val failures: List<L
+                    is ExpiryTooSoon -> PartFailure.TemporaryRemoteFailure
+                    ExpiryTooFar -> PartFailure.TemporaryRemoteFailure
+                    is ChannelDisabled -> PartFailure.TemporaryRemoteFailure
+                    is TemporaryChannelFailure -> PartFailure.TemporaryRemoteFailure
+                    TemporaryNodeFailure -> PartFailure.TemporaryRemoteFailure
+                    PermanentChannelFailure -> PartFailure.TemporaryRemoteFailure
+                    PermanentNodeFailure -> PartFailure.TemporaryRemoteFailure
+                    is InvalidOnionBlinding -> PartFailure.Uninterpretable(failure.value.message)
+                    is InvalidOnionHmac -> PartFailure.Uninterpretable(failure.value.message)
+                    is InvalidOnionKey -> PartFailure.Uninterpretable(failure.value.message)
+                    is InvalidOnionPayload -> PartFailure.Uninterpretable(failure.value.message)
+                    is InvalidOnionVersion -> PartFailure.Uninterpretable(failure.value.message)
+                    InvalidRealm -> PartFailure.Uninterpretable(failure.value.message)
+                    is IncorrectCltvExpiry -> PartFailure.Uninterpretable(failure.value.message)
+                    RequiredChannelFeatureMissing -> PartFailure.Uninterpretable(failure.value.message)
+                    RequiredNodeFeatureMissing -> PartFailure.Uninterpretable(failure.value.message)
+                    is UnknownFailureMessage -> PartFailure.Uninterpretable(failure.value.message)
+                }
+            }
+            return LightningOutgoingPayment.Part.Status.Failed(converted, completedAt)
         }
 
-        fun isRouteError(failure: LightningOutgoingPayment.Part.Status.Failed) = when (failure.remoteFailureCode) {
-            UnknownNextPeer.code -> true
-            ChannelDisabled.code -> true
-            TemporaryChannelFailure.code -> true
-            PermanentChannelFailure.code -> true
-            TemporaryNodeFailure.code -> true
-            PermanentNodeFailure.code -> true
+        /** Return true if this is a remote routing error. */
+        fun isRouteError(failure: LightningOutgoingPayment.Part.Status.Failed) = when (failure.failure) {
+            is PartFailure.TemporaryRemoteFailure -> true
+            is PartFailure.RecipientLiquidityIssue -> true
+            is PartFailure.RecipientIsOffline -> true
             else -> false
         }
 
-        fun isRejectedByRecipient(failure: LightningOutgoingPayment.Part.Status.Failed) = when (failure.remoteFailureCode) {
-            IncorrectOrUnknownPaymentDetails.code -> true
+        /** Return true if the recipient received the payment and rejected it, indicating that we shouldn't retry. */
+        fun isRejectedByRecipient(failure: LightningOutgoingPayment.Part.Status.Failed) = when (failure.failure) {
+            is PartFailure.RecipientRejectedPayment -> true
             else -> false
         }
```

Comment:
This is so straightforward we can probably remove this function entirely?

```suggestion
        /** Return true if the recipient received the payment and rejected it, indicating that we shouldn't retry. */
        fun isRejectedByRecipient(failure: LightningOutgoingPayment.Part.Status.Failed) = failure.failure is PartFailure.RecipientRejectedPayment
```

Reply:
Done in https://github.com/ACINQ/lightning-kmp/pull/634/commits/ea4eb74b9fe11277ccf93a28ebb249820b322374